rust-install/src/dist.rs

```diff
@@ -268,3 +273,84 @@ pub fn download_hash(url: &str, cfg: DownloadCfg) -> Result<String> {
 
     Ok(try!(utils::read_file("hash", &hash_file).map(|s| s[0..64].to_owned())))
 }
+
+// Installs or updates a toolchain from a dist server. If an initial
+// install then it will be installed with the default components. If
+// an upgrade then all the existing components will be upgraded.
+//
+// Returns the manifest's hash if anything changed.
+pub fn update_from_dist<'a>(download: DownloadCfg<'a>,
+                            update_hash: Option<&Path>,
+                            toolchain: &str,
+                            prefix: &InstallPrefix,
+                            add: &[Component],
+                            remove: &[Component],
+                            ) -> Result<Option<String>> {
+
+    let ref toolchain = try!(ToolchainDesc::from_str(toolchain).ok_or(Error::InvalidToolchainName));
+    let trip = try!(toolchain.target_triple().ok_or_else(|| Error::UnsupportedHost(toolchain.full_spec())));
+
+    let manifestation = try!(Manifestation::open(prefix.clone(), &trip));
+
+    let changes = Changes {
+        add_extensions: add.to_owned(),
+        remove_extensions: remove.to_owned(),
+    };
+
+    match dl_v2_manifest(download, update_hash, toolchain) {
```

Comment:
Would be nice to have a notification for which type of manifest is going to be used, (can just add a TODO for now)